File: medium/22. Generate Parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        backtracking(result, "", 0, 0, n);
        return result;
    }

    void backtracking (List<String> result, String str, int s, int e, int max) {
        if(str.length() == max*2) {
            result.add(str);
            return;
        }
        
        if(s<max) backtracking(result, str + "(", s+1, e, max);
        if(e<s) backtracking(result, str + ")", s, e+1, max);
    }
}

File: medium/22. Generate Parentheses.py
class Solution:
    #TC : O(2^2N) = O(4^n) -> assume we have 2 options and it is repeated 2n times
    def generateParenthesis(self, n: int) -> List[str]:
        self.res = []

        def dfs(n, open, close, path) :
            #print(path)
            if len(path) == n*2 :
                self.res.append(path[:])
                return

            if open < n :
                path = ''.join((path,'('))
                dfs(n, open+1, close, path)
                path = path[:-1]
            
            if close < open :
                path = ''.join((path,')'))
                dfs(n, open, close+1, path)
                path = path[:-1]

        dfs(n, 0, 0, "")
        return self.res
        
            

        
